docs/topics/tasks.txt

```diff
@@ -0,0 +1,337 @@
+
+Backend arguments
+-----------------
+
+Each backend can be given additional arguments to control its behavior.
+These arguments are provided as additional keys in the :setting:`TASKS` setting.
+Valid arguments are as follows:
+
+* :setting:`ENQUEUE_ON_COMMIT <TASKS-ENQUEUE_ON_COMMIT>`: Whether a task should
+  be enqueued at the end of the current transaction (if there
+  is one) commits successfully, rather than enqueueing immediately. This argument
+  defaults to ``True``.
+
+* :setting:`QUEUES <TASKS-QUEUES>`: Restrict the queue names a task may be
+  enqueued to. By default, only the ``"default"`` queue is allowed. Queue name
+  validation can be disabled by setting this to an empty list.
+
+Asynchronous support
+--------------------
+
+Django has developing support for asynchronous task backends.
+
+``django.tasks.backends.base.BaseTaskBackend`` has async variants of all base
+methods. By convention, the asynchronous versions of all methods are prefixed
+with ``a``. The arguments for both variants are the same.
+
+Retrieving backends
+-------------------
+
+Backends can be retrieved using the ``tasks`` connection handler::
+
+    from django.tasks import tasks
+
+    tasks["default"]  # The default backend
+    tasks["reserve"]  # Another backend
+
+The "default" backend is available as ``default_task_backend``::
+
+    from django.tasks import default_task_backend
+
+Backend features
+----------------
+
+Django's tasks framework is designed to be backend-agnostic. This means some
+backends may not implement the same features as others, such as executing ``async``
+functions.
+
+Depending on the backend configured, it may be necessary to gracefully degrade
+functionality, or use a system check to prevent the application from starting at all.
+
+To facilitate this, certain features can be checked on a backend:
+
+* :attr:`supports_defer <django.tasks.backends.base.BaseTaskBackend.supports_defer>`:
+  Can tasks be executed after a specific time using ``run_after``?
+
+* :attr:`supports_async_task <django.tasks.backends.base.BaseTaskBackend.supports_async_task>`:
+  Can ``async`` functions (coroutines) be used as task functions?
+
+* :attr:`supports_get_result <django.tasks.backends.base.BaseTaskBackend.supports_get_result>`:
+  Can a task's results be retrieved from another thread or process?
+
+Defining tasks
+==============
+
+Tasks are defined using the :meth:`django.tasks.task` decorator on a
+module-level function::
+
+    from django.tasks import task
+
+
+    @task()
+    def calculate_meaning_of_life(answer=42):
+        return answer
+
+Returned in a :class:`django.tasks.Task` instance.
+
+.. note::
+
+    All arguments and return values must be JSON serializable, and round-trip
+    as the same types.
+
+    This means complex types like model instances, as well as many built-in types
+    like ``datetime`` and ``tuple`` cannot be used in tasks without additional
+    conversion.
```

Comment:
Oh you're totally right.

So a bit of a variant on this idea: right now the task backend handles the call to `json_normalize`. What about changing this so that the backends expect already serialized args/kwargs? And `Task` has a method like `serialize_arguments` which returns a bundle meant to be provided to the backend?

That way the backends are not in charge of the normalization step, and a project could easily subclass `Task` to add things that they care about and do their own serialization/deserialization strategy if they care.

That way we enforce the JSON-ness of the task-provided arguments at the backend API level, whereas in the existing design I would be tempted to subclass a backend to provide different serialization code. Tasks could remain backend-agnostic.

Not a blocker, but I do think if we're worried about backend-specific behavior then having the backend class _not_ be in charge of JSON normalization would help avoid that kind of issue.